src/code/z_demo.c

```diff
@@ -971,144 +1003,146 @@ s32 Cutscene_CountNormalMasks(void) {
-                        } else {
-                            func_801477B4(play);
-                            csCtx->frames++;
                         }
+                    } else {
+                        func_801477B4(play);
+                        csCtx->curFrame++;
                     }
                 }
-
-                if ((talkState == TEXT_STATE_5) && Message_ShouldAdvance(play)) {
-                    func_80152434(play, cmd->base);
-                }
             }
 
-            if ((talkState == TEXT_STATE_CLOSING) && (D_801BB160 == CS_TEXTBOX_TYPE_3)) {
-                csCtx->frames--;
-                D_801BB124++;
+            if ((talkState == TEXT_STATE_5) && Message_ShouldAdvance(play)) {
+                func_80152434(play, cmd->textId);
             }
+        }
 
-            if (originalCsFrames == csCtx->frames) {
-                Interface_SetHudVisibility(HUD_VISIBILITY_NONE);
-                D_801BB124 = 0;
-                D_801BB128 = 0;
-                func_80161C0C();
-            } else {
-                func_80161BE0(1);
-            }
+        if ((talkState == TEXT_STATE_CLOSING) && (sCutsceneTextboxType == CS_TEXT_TYPE_3)) {
+            csCtx->curFrame--;
+            sCurTextId++;
+        }
+
+        if (endFrame == csCtx->curFrame) {
+            Interface_SetHudVisibility(HUD_VISIBILITY_NONE);
+            sCurTextId = 0;
+            sCurOcarinaAction = 0;
+            CutsceneCamera_Reset();
+        } else {
+            CutsceneCamera_SetState(CS_CAM_STATE_UPDATE_SPLINE);
         }
     }
+end:;
 }
 
-// Related to actorActions. Maybe a generic actorAction setter?
-void func_800ECD7C(CutsceneContext* csCtx, u8** cutscenePtr, s16 index) {
+void Cutscene_SetActorCue(CutsceneContext* csCtx, u8** script, s16 cueChannel) {
     s32 i;
-    s32 count;
+    s32 numCues;
 
-    bcopy(*cutscenePtr, &count, sizeof(s32));
-    *cutscenePtr += sizeof(s32);
+    bcopy(*script, &numCues, sizeof(numCues));
+    *script += sizeof(numCues);
 
-    for (i = 0; i < count; i++) {
-        CsCmdActorAction* actorAction = *(CsCmdActorAction**)cutscenePtr;
+    for (i = 0; i < numCues; i++) {
+        CsCmdActorCue* cue = *(CsCmdActorCue**)script;
 
-        if ((csCtx->frames >= actorAction->startFrame) && (csCtx->frames < actorAction->endFrame)) {
-            csCtx->actorActions[index] = actorAction;
+        if ((csCtx->curFrame >= cue->startFrame) && (csCtx->curFrame < cue->endFrame)) {
+            csCtx->actorCues[cueChannel] = cue;
         }
 
-        *cutscenePtr += sizeof(CsCmdActorAction);
+        *script += sizeof(CsCmdActorCue);
     }
 }
 
 /**
- * Loops over the cutscene data itself (`cutscenePtr`), applying the effects of each command instantaneously (for most
+ * Loops over the cutscene data itself (`script`), applying the effects of each command instantaneously (for most
  * commands).
  *
  * The cutscene data is an irregularly-structured array of words, which is made up of
```

Comment:
also not really related to these changes, but I find alot of the things this large comment says to either not be needed, or probably not be in the best place. 

when I have more time to think about it, I can possibly suggest some rewording to cut down on things while still giving the important points